### ISSUE TITLE
Deserialized arrays with base mutate strings (migrated from Trac #1233)

### ISSUE THREAD

**Issue by thouis** [thouis/numpy-trac-migration#2786]
Original ticket http://projects.scipy.org/numpy/ticket/1233
Reported 2009-09-23 by trac user hniksic, assigned to atmention:mwiebe.

Numpy arrays with the "base" property are deserialized as arrays pointing to a storage contained within a Python string.  This is a problem since such arrays are mutable and can mutate existing strings.  Here is how to create one:

```
>>> import numpy, cPickle as p
>>> a = numpy.array([1, 2, 3])   # create an array
>>> b = a[::-1]                  # create a view
>>> b
array([3, 2, 1])
>>> b.base                       # view's base is the original array
array([1, 2, 3])
>>> c = p.loads(p.dumps(b, -1))  # roundtrip the view through pickle
>>> c
array([3, 2, 1])
>>> c.base                       # base is now a simple string:
'\x03\x00\x00\x00\x02\x00\x00\x00\x01\x00\x00\x00'
>>> s = c.base
>>> s
'\x03\x00\x00\x00\x02\x00\x00\x00\x01\x00\x00\x00'
>>> type(s)
<type 'str'>
>>> c[0] = 4                     # when the array is mutated...
>>> s                            # ...the string changes value!
'\x04\x00\x00\x00\x02\x00\x00\x00\x01\x00\x00\x00'
```

This is somewhat disconcerting, as Python strings are supposed to be immutable.  In this case the string was created by numpy and is probably not shared by anyone, so it doesn't present a problem in practice.  But in corner cases it can lead to serious bugs.  Python has a cache of one-character strings, which cannot be turned off.  This means that one-byte array views can change existing Python strings used elsewhere in the code.  For example:

```
>>> a = numpy.array([65], 'int8')
>>> b = a[::-1]
>>> c = p.loads(p.dumps(b, -1))
>>> c
array([65], dtype=int8)
>>> c.base
'A'
>>> c[0] = 66
>>> c.base
'B'
>>> 'A'
'B'
```

Note how changing a numpy array permanently changed the contents of all 'A' strings in this python instance, rendering python unusable.

The fix should be straightforward: use a string subclass (which will
skip the one-letter cache), or an entirely separate type for storage of "base" memory referenced by deserialized arrays.

**Comment by thouis** [comment]
Comment in Trac by atmention:mwiebe, 2011-03-24

The first case no longer creates a view into a string in 1.6 beta, but the second case still does. The unpickled array should always own its data.

The behavior of Python's internal 'A' string being changed is still there:

```
In [64]: a = numpy.array([65], 'int8')

In [65]: b = a[::-1]

In [66]: c = p.loads(p.dumps(b, -1))

In [67]: c
Out[67]: array([65], dtype=int8)

In [68]: c.base
Out[68]: 'A'

In [69]: c[0] = 66

In [70]: c.base
Out[70]: 'B'

In [71]: type(c.base)
Out[71]: str

In [72]: 'A'
Out[72]: 'B'
```

**Comment by thouis** [comment]
Comment in Trac by atmention:rgommers, 2011-03-30

Mark, if this has to go into 1.6.0, can you find time to fix it in the next couple of weeks?

**Comment by thouis** [comment]
Comment in Trac by atmention:pv, 2011-04-02

Addressed in commit:7b39b9e5  (same as #1708)

This can be fixed without giving up essentially any speed advantage by copying data if the string happens to be interned. Interned strings are typically always short, so for long strings no copy is made.

Yes, this is an "evil" approach that uses some knowledge on how CPython works. However, a speed advantage here is a bit questionable.

**Comment by thouis** [comment]
Comment in Trac by atmention:pv, 2011-04-02

a bit questionable .... so the whole approach needs probably be revised at some point. However, the implementation should be "safe" now.
